Editorial review: Add moveBefore() docs (#39002)
mdn/content: files/en-us/web/api/document/movebefore/index.md

There are some constraints to be aware of when using `moveBefore()`:

- It can only work when moving a node within the same document.
- It won't work if you try to move a node that is not connected to the DOM to an already connected parent, or vice versa.

In such cases, `moveBefore()` will fail with a `HierarchyRequestError` exception. If the above constraints are requirements for your particular use case, you should use {{domxref("Node.insertBefore()")}} instead, or use [`try...catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) to handle the errors that arise from such cases.

## Examples

### Moving a comment node with `moveBefore()`

In this demo we show how to use `document.moveBefore()` to move a comment node within the DOM.

#### HTML

The HTML is a minimal template that features a comment inside the {{htmlelement("body")}}.

```html live-sample___movebefore-comment
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>document.moveBefore() example</title>
  </head>
  <body>
    <!-- This comment should be at the end of the document -->
    <p>Some content</p>
  </body>
</html>
```

#### JavaScript

> In our script, we loop through all the {{domxref("Node.childNodes", "childNodes")}} of the `<body>` element. When we find a node with a {{domxref("Node.nodeType", "nodeType")}} value of `8` (which indicates a comment node), we store a reference to it insidea variable called `commentNode`. We then invoke `document.moveBefore()`, specifying that we want to move the comment node, and specifying a second argument of `null` to inserted our comment the end of the `Document`'s child nodes.

Reviewer: ```suggestion
In our script, we loop through all the {{domxref("Node.childNodes", "childNodes")}} of the `<body>` element. When we find a node with a {{domxref("Node.nodeType", "nodeType")}} value of `8` (which indicates a comment node), we store a reference to it inside a variable called `commentNode`. We then invoke `document.moveBefore()`, specifying that we want to move the comment node, and specifying a second argument of `null` to insert our comment the end of the `Document`'s child nodes.
```

---
title: "Document: moveBefore() method"
short-title: moveBefore()
slug: Web/API/Document/moveBefore
page-type: web-api-instance-method
browser-compat: api.Document.moveBefore
---

{{APIRef("DOM")}}

The **`moveBefore()`** method of the {{domxref("Document")}} interface moves a given {{domxref("Node")}} inside the `Document` DOM node as a direct child, before a given reference node.

## Syntax

```js-nolint
moveBefore(movedNode, referenceNode)
```

### Parameters

- `movedNode`

>   - : A {{domxref("Node")}} representing the node to be moved.

Reviewer: ```suggestion
  - : A {{domxref("Node")}} representing the node to be moved. Note that this must be an {{domxref("Element")}} or a {{domxref("CharacterData")}} node.
```

Because I would be annoyed if I tried to use something else, which was correct according to the parameter description, and it didn't work.

- `referenceNode`
  - : A {{domxref("Node")}} that `movedNode` will be moved before, or `null`. If the value is `null`, `movedNode` is inserted at the end of the `Document`'s child nodes.

### Return value

None ({{jsxref("undefined")}}).

### Exceptions

- `HierarchyRequestError` {{jsxref("TypeError")}}
  - : Thrown in any of the following situations:
    - The specified `movedNode` is not part of this document.
    - The specified `movedNode` is not an {{domxref("Element")}} or {{domxref("CharacterData")}} node.

>     - You are trying to move `movedNode` before the document's [Doctype](/en-US/docs/Glossary/Doctype) (represented by a {{domxref("DocumentType")}} object).

Reviewer: ```suggestion
    - You are trying to move `movedNode` before the document's {{glossary("doctype")}} (represented by a {{domxref("DocumentType")}} object).
```

(note that this might give you a flaw because the case of "doctype" does not match the slug case, but I think we ought to ignore that).

---
title: "Document: moveBefore() method"
short-title: moveBefore()
slug: Web/API/Document/moveBefore
page-type: web-api-instance-method
browser-compat: api.Document.moveBefore
---

{{APIRef("DOM")}}

The **`moveBefore()`** method of the {{domxref("Document")}} interface moves a given {{domxref("Node")}} inside the `Document` DOM node as a direct child, before a given reference node.

## Syntax

```js-nolint
moveBefore(movedNode, referenceNode)
```

### Parameters

- `movedNode`
  - : A {{domxref("Node")}} representing the node to be moved.
- `referenceNode`
  - : A {{domxref("Node")}} that `movedNode` will be moved before, or `null`. If the value is `null`, `movedNode` is inserted at the end of the `Document`'s child nodes.
    > [!NOTE]

>     > The `referenceNode` parameter is mandatory. You must explicitly pass a {{domxref("Node")}} or `null`.

Author: This is a good point; I wouldn't normally add this. I only did it because `insertBefore()` has a similar note on it.

I have removed all three instances from this PR (but not the `insertBefore()` one, as that has additional bug information that I'm not sure about removing.)

### Return value

> None ({{jsxref("Undefined")}}).

Author: done

### Exceptions

- `HierarchyRequestError` {{jsxref("TypeError")}}

>   - : Thrown when:

Author: updated, on all three pages

>     - The specified `referenceNode` is not yet added to the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

Author: I think the confusion here was partly around `referenceDOM`, which should have been `movedNode`, I think. I've updated the sentence in both places it appears, to:

> The specified `movedNode` is not part of the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

Does this clear up the confusion, or no? In which case, please state your case again ;-)

>     - The specified `referenceNode` is not yet added to the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

Reviewer: I no longer understand my own previous comment (!) but:

> The specified movedNode is not part of the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

This reads weirdly for `**Document**.moveBefore()`, because with that API, per the overview description, we are moving "a given [Node](https://pr39002.review.mdn.allizom.net/en-US/docs/Web/API/Node) inside the Document DOM node as a direct child".

Could we replace both bullets with "The specified `movedNode` is not part of this document."

Also there seem to be some other conditions in which this exception gets thrown: https://dom.spec.whatwg.org/#move.

>     - The specified `referenceNode` is not yet added to the DOM, and you are trying to move it inside a node that is part of the DOM, or vice versa.

Author: OK, I've made the suggested change. I've also looked through the spec and updated the situations under which this exception gets thrown.

    - An attempt is made to move `movedNode` between two different documents.
- `NotFoundError` {{jsxref("TypeError")}}
  - : The specified `referenceNode` is not a child of the node you are calling `moveBefore()` on, that is, the node you are trying to move `movedNode` inside.
- `TypeError` {{jsxref("TypeError")}}
  - : The second argument was not supplied.

## Description

The `moveBefore()` method moves a given node to a new place in the DOM. It provides similar functionality to the {{domxref("Node.insertBefore()")}} method, except that it doesn't remove and then reinsert the node. This means that the state of the node (which would be reset if moving it with `insertBefore()` and similar mechanisms) is preserved after the move. This includes:

- [Animation](/en-US/docs/Web/CSS/CSS_animations) and [transition](/en-US/docs/Web/CSS/CSS_transitions) state.
- {{htmlelement("iframe")}} loading state.

> - Interactive node {{cssxref(":focus")}} and {{cssxref(":active")}} states.

Reviewer 2: Perhaps reword to "interactivity"

Author: Updated to

> - Interactivity states (for example, {{cssxref(":focus")}} and {{cssxref(":active")}}).

- [Fullscreen](/en-US/docs/Web/API/Fullscreen_API) element state.
- Open/close state of [popovers](/en-US/docs/Web/API/Popover_API).

> - Modal state of {{htmlelement("dialog")}} elements (modal dialog will not be closed).

Author: Good spot, thanks! I've fixed it in all three places.

The play state of {{htmlelement("video")}} and {{htmlelement("audio")}} elements is not included in the above list, as these elements retain their state when removed and reinserted, regardless of the mechanism used.

> [!NOTE]

> > When observing changes to the DOM using a {{domxref("MutationObserver")}}, nodes moved with `moveBefore()` will be recorded with a [removed node](/en-US/docs/Web/API/MutationRecord/removedNodes) and an [added node](/en-US/docs/Web/API/MutationRecord/addedNodes).

Author: OK, WFM. I've moved it out of a note, in all three pages.

### `moveBefore()` constraints

There are some constraints to be aware of when using `moveBefore()`:

- It can only work when moving a node within the same document.
- It won't work if you try to move a node that is not connected to the DOM to an already connected parent, or vice versa.

In such cases, `moveBefore()` will fail with a `HierarchyRequestError` exception. If the above constraints are requirements for your particular use case, you should use {{domxref("Node.insertBefore()")}} instead, or use [`try...catch`](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) to handle the errors that arise from such cases.

## Examples

> The `moveBefore()` method is not particularly useful when invoked on the `Document` node. There are some non-element uses for it, for example you could use `moveBefore()` to move comment nodes around the root of the `Document`. However, you are much more likely to find a use for invoking it on an individual {{domxref("Element")}} — see {{DOMxRef("Element.moveBefore()")}} for some examples.

Author: I've moved the paragraph to the description section.

And, I've added a demo that moves a comment node to the end of the document. It's pretty pointless, and not something you'd ever really do, but it's about the best I can come up with (I previously spent hours trying to even get `document.moveBefore()` to run, but now I have ... hurrah!)
